### DISCUSSION TITLE
Trying to understand how to use JSX correctly with Hastscript and hast-util-to-jsx-runtime

### DISCUSSION THREAD

**Discussion by abeeonthemoon** [remarkjs/remark#1149] [Q&A]
Hi everyone, I am relatively new to remark and I am trying to achieve the following:

1) Create a custom remark-directive
2) Map the custom remark-directive to a React Component
3) Correctly convert the markdown to JSX

In trying to map the mdast tree to a hast tree, I have come to the understanding that I need to understand how to use hastscript and hast-util-to-jsx-runtime. 

I have prepared a small example that demonstrates my understanding of hastscript and hast-util-to-jsx-runtime in relation to this problem.

```javascript
import { read } from 'to-vfile';
import { unified } from 'unified';
import remarkParse from 'remark-parse';
import remarkDirective from 'remark-directive';
import remarkRehype from 'remark-rehype';
import rehypeFormat from 'rehype-format';
import rehypeStringify from 'rehype-stringify';
import { visit } from 'unist-util-visit';
import { h } from 'hastscript';
import { toJsxRuntime } from 'hast-util-to-jsx-runtime';
import { Fragment, jsx, jsxs } from 'react/jsx-runtime';

const sampleText = `
# Hello

:::sampleDirective

This is the child text of sample component.

:::

`;

const SampleComponent = ({ valueText }) => {
  return (
    <div>
      <p>{props.valueText}</p>
    </div>
  );
};

runProcessor();

export const runProcessor = async () => {

  const file = await unified()
    .use(remarkParse)
    .use(remarkDirective)
    .use(myRemarkPlugin)
    .use(remarkRehype)
    .use(myJsxPlugin)
    .use(rehypeFormat)
    .use(rehypeStringify)
    .process(sampleText);


  console.log('@@@ file', file);
};

// This plugin is an example to let users write HTML with directives.
// It’s informative but rather useless.
// See below for others examples.
/** @type {import('unified').Plugin<[], import('mdast').Root>} */
function myRemarkPlugin() {
  return (tree) => {
    visit(tree, (node) => {
      if (
        node.type === 'textDirective' ||
        node.type === 'leafDirective' ||
        node.type === 'containerDirective'
      ) {
        const data = node.data || (node.data = {});
        const hast = h(node.name, {
          valueText: 'This is an alternative text being passed to SampleComponent',
        });

        data.hName = hast.tagName;
        data.hProperties = hast.properties;
      }
    });
  };
}

function myJsxPlugin() {
  return (tree) => {

    const result = toJsxRuntime(tree, {
      Fragment,
      jsx,
      jsxs,
      components: {
        sampledirective: (props) => <SampleComponent {...props} />,
      },
    });

  };
}
```

What I am aiming for this to print is:

```jsx
<h1>hello</h1>
<SampleComponent valueText="This is a sample component">
  <p>What could it be</p>
</SampleComponent>
```

But what it's actually printing is:

```jsx
<h1>hello</h1>
<sampledirective valueText="This is a sample component">
  <p>What could it be</p>
</sampledirective>
```

This means that at no point do any of the plugins correctly convert a node into a jsx element. How can I use hastscript or hast-util-to-jsx-runtime to process JSX and React Components? I have read the documentations for both libraries, as well as this issue https://github.com/syntax-tree/hastscript/issues/17: However I am still confused.

Please let me know if I should clarify anything, I would love to understand these libraries and their use cases better.

**Comment by wooorm** [comment]
Hey!

> Create a custom `remark-directive`

Why use directives when you want JSX? Why not write JSX instead? They solve the same problem: components in markdown.

> Map the custom `remark-directive` to a React Component

Depending on what you want, you might not need to.
You can pass `components` to `toJsxRuntime` in `options`: https://github.com/syntax-tree/hast-util-to-jsx-runtime/tree/main#options.
so you can instead turn directives into a “normal” HTML custom element (something with a tag name that includes a `-`). And then handle it in `components`: pass `'my-element': MyComponent` to turn your `my-element`, which was a directive, into `MyComponent`.

> Correctly convert the markdown to JSX

Assuming you do instead actually need to generate JSX, which is JavaScript code that still needs to be evaluated somewhere, you are missing two pieces:

1. https://github.com/syntax-tree/hast-util-to-estree
2. you still need to **evaluate**, in otherwords, **run**, the code somewhere.
   Either by writing to a file and then importing it, or by calling `eval()` (or `new Function`).

However, by now, you have invented MDX. Which is a fun learning exercise, but I would instead recommend using the existing project.

You would still need to support directives in MDX, which you could do with a remark plugin that creates a JSX node:

*   For the AST of directives, see: https://github.com/syntax-tree/mdast-util-directive#syntax-tree
*   For the AST of JSX, see: https://github.com/syntax-tree/mdast-util-mdx-jsx#syntax-tree

---

In the rest of your text, you are confusing what the terms JSX and components mean. JSX is a syntax. It’s an extension to a programming language, a file format, which still needs to be compiled and evaluated. Components are reusable things.
anyway, I think my above points will help you forwards!

**Reply by abeeonthemoon** [reply]
Hi @wooorm thank you so much for your response, I really appreciate it.

> You can pass components to toJsxRuntime in options: https://github.com/syntax-tree/hast-util-to-jsx-runtime/tree/main#options.

I have been able to make this work and `toJsxRuntime` returns a JSX.Element that I am able to render in my React App.

However, I have some follow-up questions I'd like to ask.

### Question 1

Can I preserve my custom component in the hast tree?

I am wondering if it's possible to maintain my component in the hast tree using hastscript or some other library. In the hastscript documentation, under the JSX section it states:

> This package can be used with JSX. You should use the automatic JSX runtime set to `hastscript` (also available as the more explicit name `hastscript/html` or `hastscript/svg`.

and then it provides the following example:

```jsx
/** @jsxImportSource hastscript */
console.log(
  <div class="foo" id="some-id">
    <span>some text</span>
    <input type="text" value="foo" />
    <a class="alpha bravo charlie" download>
      deltaecho
    </a>
  </div>
)
```

I am unclear on how or where I set JSX runtime to 'hastscript' and how the JSX syntax in the example is passed into the hast tree. I am trying to see if I can stringify the hast tree into the following snippet from the above example - where the markdown directive is replaced with the custom component:

```jsx
<h1>hello</h1>
<SampleComponent valueText="This is a sample component">
  <p>What could it be</p>
</SampleComponent>
```

I also think understanding this more clearly would help me understand this issue: https://github.com/syntax-tree/hastscript/issues/17. Because there, the way in which hastscript is being used with JSX is just unclear to me. 

### Question 2

> Assuming you do instead actually need to generate JSX, which is JavaScript code that still needs to be evaluated somewhere, you are missing two pieces:
>  1. https://github.com/syntax-tree/hast-util-to-estree
>  2. you still need to evaluate, in otherwords, run, the code somewhere.
Either by writing to a file and then importing it, or by calling eval() (or new Function).

Going back to my example, why do I need to build an estree when `toJsxRuntime` returns a JSX element? As I understand, I can just use the newly created JSX.Element in my app. I realize the "why" might be a big question - but is it some kind of good practice or concept that it adheres to.

> However, by now, you have invented MDX. Which is a fun learning exercise, but I would instead recommend using the existing project.

Yes, I am actually just trying to learn this ecosystem and get a better grasp on things.


Thanks again!

**Reply by wooorm** [reply]
> Can I preserve my custom component in the hast tree?

Why?
Do you mean the component, as in a function or class?
Or do you mean a “*called*” component? Such as a `ReactNode`?

> In the hastscript documentation, under the JSX section it states:

That section is not related to what you want.
JSX is a *syntax* that translates to function calls.
It can be used to create React, Preact, or Vue calls.
It can *also* be used to create our syntax tree, by transforming to `h` calls.
You don’t need this.

>  why do I need to build an estree when `toJsxRuntime` returns a JSX element?

I don’t yet know what you want.
If you really want *JSX*, you need to generate a string of JavaScript, and then later evaluate it. This is where estrees are used.
If you actually want components, react nodes, then JSX is unrelated, and estree is too.
